### PR TITLE
new resource aws_codebuild_project

### PR DESCRIPTION
**Pull request by liamjbennett**
New resource for the new AWS CodeBuild service

- [x] Schema
- [x] Functionality
  - [x] Create
  - [x] Read
  - [x] Update
  - [x] Delete
- [x] Tests
  - [x] Schema Validation
  - [x] Acceptance Tests
  - [x] Running Them
- [x] Documentation

### CODE REVIEW

**Comment by liamjbennett** [comment]
```
vagrant@terraform:/opt/gopath/src/github.com/hashicorp/terraform$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSCodeBuildProject_*'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/01/31 16:20:01 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSCodeBuildProject_* -timeout 120m
=== RUN   TestAccAWSCodeBuildProject_basic
--- PASS: TestAccAWSCodeBuildProject_basic (37.77s)
=== RUN   TestAccAWSCodeBuildProject_artifactsTypeValidation
--- PASS: TestAccAWSCodeBuildProject_artifactsTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_artifactsNamespaceTypeValidation
--- PASS: TestAccAWSCodeBuildProject_artifactsNamespaceTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_nameValidation
--- PASS: TestAccAWSCodeBuildProject_nameValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_descriptionValidation
--- PASS: TestAccAWSCodeBuildProject_descriptionValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_environmentComputeTypeValidation
--- PASS: TestAccAWSCodeBuildProject_environmentComputeTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_environmentTypeValidation
--- PASS: TestAccAWSCodeBuildProject_environmentTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_sourceTypeValidation
--- PASS: TestAccAWSCodeBuildProject_sourceTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_sourceAuthTypeValidation
--- PASS: TestAccAWSCodeBuildProject_sourceAuthTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_timeoutValidation
--- PASS: TestAccAWSCodeBuildProject_timeoutValidation (0.00s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    37.792s
```

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsCodeBuildProjectCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
+
+	log.Printf("[DEBUG] CodeBuild Create Project: %s", d.Id())
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: d.ID will fail here - there is no id at this point

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+
+		return resource.NonRetryableError(err)
+	})
+
+	if err != nil {
+		return fmt.Errorf("[ERROR] Error creating CodeBuild project: %s", err)
+	}
+
+	if resp.Project.Name == nil {
+		return fmt.Errorf("[ERROR] Project name was nil")
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: is this ever a possibility?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+	projectArtifacts := codebuild.ProjectArtifacts{}
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		projectArtifacts = codebuild.ProjectArtifacts{
+			Type: aws.String(data["type"].(string)),
+		}
+
+		if len(data["location"].(string)) > 0 {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: string > 0 ? - why not != "" ?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+			projectArtifacts.Location = aws.String(data["location"].(string))
+		}
+
+		if len(data["name"].(string)) > 0 {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: string > 0 ? - why not != "" ?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+			projectArtifacts.Name = aws.String(data["name"].(string))
+		}
+
+		if len(data["namespace_type"].(string)) > 0 {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: string > 0 ? - why not != "" ?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 6]
```diff
@@ -0,0 +1,722 @@
...
+			projectArtifacts.NamespaceType = aws.String(data["namespace_type"].(string))
+		}
+
+		if len(data["packaging"].(string)) > 0 {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: string > 0 ? - why not != "" ?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+			projectArtifacts.Packaging = aws.String(data["packaging"].(string))
+		}
+
+		if len(data["path"].(string)) > 0 {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: string > 0 ? - why not != "" ?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 8]
```diff
@@ -0,0 +1,722 @@
...
+	d.SetId(*resp.Project.Name)
+
+	return resourceAwsCodeBuildProjectUpdate(d, meta)
+}
+
+func expandProjectArtifacts(d *schema.ResourceData) codebuild.ProjectArtifacts {
+	configs := d.Get("artifacts").(*schema.Set).List()
+	projectArtifacts := codebuild.ProjectArtifacts{}
+
+	for _, configRaw := range configs {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: IF artifacts has MaxItems of 1, I don't think we need to range here. We can use configs[0] and grab the items directly

Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: If it doesn't have a MxItems of 1, then we need to return a []codebuild.ProjectArtifacts

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 9]
```diff
@@ -0,0 +1,722 @@
...
+}
+
+func expandProjectEnvironment(d *schema.ResourceData) *codebuild.ProjectEnvironment {
+	configs := d.Get("environment").(*schema.Set).List()
+	projectEnv := &codebuild.ProjectEnvironment{}
+
+	envConfig := configs[0].(map[string]interface{})
+
+	if v := envConfig["compute_type"]; v != nil {
+		computeType := v.(string)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: these lines can be combined as follows:

```
if v := envConfig["compute_type"]; v != nil {
    projectEnv.ComputeType = aws.String(v.(string))
}
```

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 10]
```diff
@@ -0,0 +1,722 @@
...
+
+func resourceAwsCodeBuildProject() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCodeBuildProjectCreate,
+		Read:   resourceAwsCodeBuildProjectRead,
+		Update: resourceAwsCodeBuildProjectUpdate,
+		Delete: resourceAwsCodeBuildProjectDelete,
+
+		Schema: map[string]*schema.Schema{
+			"artifacts": &schema.Schema{
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: I think this Set may need a hash

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 11]
```diff
@@ -0,0 +1,722 @@
...
+	}
+
+	// if nothing was found, then return no state
+	if len(resp.Projects) == 0 {
+		log.Printf("[INFO]: No projects were found, removing from state")
+		d.SetId("")
+		return nil
+	}
+
+	project := resp.Projects[0]
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: Can we guarantee it's always Project[0]? Should we range and check on the project name?

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 12]
```diff
@@ -0,0 +1,722 @@
...
+
+	artifacts := flattenAwsCodebuildProjectArtifacts(project.Artifacts)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: I think this should follow the if err :- d.Set pattern from environment and source

File: builtin/providers/aws/resource_aws_codebuild_project.go [hunk 13]
```diff
@@ -0,0 +1,722 @@
...
+
+	if err := d.Set("tags", tagsToMapCodeBuild(project.Tags)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAwsCodeBuildProjectUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: Do we want to update only the pieces that need to change or pass the entire UpdateProjectInput every time? In an Update func, we usually only send the pieces that change by using the if d.HasChange("")

Can this pattern be used here too?

File: website/source/docs/providers/aws/r/codebuild_project.html.markdown [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+  tags {
+    "Environment" = "Test"
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
```
Comment (stack72) [inline website/source/docs/providers/aws/r/codebuild_project.html.markdown]: Are all of these args optional except name? The schema suggests that's not the case

File: website/source/docs/providers/aws/r/codebuild_project.html.markdown [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+* `id` - The ARN of the CodeBuild project.
+* `description` - A short description of the project.
+* `encryption_key` - The AWS Key Management Service (AWS KMS) customer master key (CMK) that was used for encrypting the build project's build output artifacts.
+* `name` - The projects name.
+* `service_role` - The ARN of the IAM service role.
+
+## Import
+
+CodeBuild projects can be imported using the ARN, e.g
```
Comment (stack72) [inline website/source/docs/providers/aws/r/codebuild_project.html.markdown]: If we are not importing, then please remove this

File: builtin/providers/aws/resource_aws_codebuild_project.go (hunk 11 above)
Comment (liamjbennett) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: Updated the Id to use the ARN not the name so the lookup here will only ever return 1 result.

File: builtin/providers/aws/resource_aws_codebuild_project.go (hunk 13 above)
Comment (liamjbennett) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: Updated to use that pattern.

File: builtin/providers/aws/resource_aws_codebuild_project.go (hunk 10 above)
Comment (liamjbennett) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: please explain?

Comment (liamjbennett) [inline builtin/providers/aws/resource_aws_codebuild_project.go]: Added `Set: resourceAwsCodeBuildProjectArtifactsHash`

**Comment by liamjbennett** [comment]
```
vagrant@terraform:/opt/gopath/src/github.com/hashicorp/terraform$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSCodeBuildProject_*'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/02 15:45:10 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSCodeBuildProject_* -timeout 120m
=== RUN   TestAccAWSCodeBuildProject_basic
--- PASS: TestAccAWSCodeBuildProject_basic (38.56s)
=== RUN   TestAccAWSCodeBuildProject_artifactsTypeValidation
--- PASS: TestAccAWSCodeBuildProject_artifactsTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_artifactsNamespaceTypeValidation
--- PASS: TestAccAWSCodeBuildProject_artifactsNamespaceTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_nameValidation
--- PASS: TestAccAWSCodeBuildProject_nameValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_descriptionValidation
--- PASS: TestAccAWSCodeBuildProject_descriptionValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_environmentComputeTypeValidation
--- PASS: TestAccAWSCodeBuildProject_environmentComputeTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_environmentTypeValidation
--- PASS: TestAccAWSCodeBuildProject_environmentTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_sourceTypeValidation
--- PASS: TestAccAWSCodeBuildProject_sourceTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_sourceAuthTypeValidation
--- PASS: TestAccAWSCodeBuildProject_sourceAuthTypeValidation (0.00s)
=== RUN   TestAccAWSCodeBuildProject_timeoutValidation
--- PASS: TestAccAWSCodeBuildProject_timeoutValidation (0.00s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    38.584s
```

**Comment by stack72** [comment]
Thanks so much @liamjbennett - this is great! This LGTM :)

**Comment by stack72** [comment]
I just manually merged this after rebasing out all of the CHANGELOG entries

**Comment by paulcollinsiii** [comment]
@stack72 I'm probably missing something, but I'm not seeing the documentation changes from this PR on master. Was it missed during the rebasing?

**Comment by stack72** [comment]
Hi @paulcollinsiii 

this has been rectified and is going to the website within the next 30 mins

p.

**Comment by jagregory** [comment]
@stack72 The website still doesn't seem to have the CodeBuild docs. I can see them in the website repo, just not published on the site itself.

**Comment by stack72** [comment]
Thanks @jagregory  - fixing now!

**Pull request closed without merging**